### ISSUE TITLE
Client error 404 while linking files

### ISSUE THREAD

**Issue by amanjeetsingh** [goldmann/docker-squash#99]
Hi,

This is similar to the issue addressed in https://github.com/goldmann/docker-squash/pull/82

I am trying to squash a very large Docker image (approximately 50 GB) which we use as part of our automated build process. Overtime, due the volume of changes that we have committed, the size of the image has grown over to 50 GB. This is slowing down the build because it takes a while to download the image etc. 

docker-squash runs for hours (which is understandable), and then it fails while linking files right at the end. docker-squash worked fine for other images so it's definitely not linked to the environment. The issue is definitely with this particular image that we are trying to squash. By the looks of it, docker-squash errors out while loading squashed image. I have attached the logs here with this issue. 

We are running Python version 2..7.10 and Docker version 1.10.2 (build c3959b1)

Any help would be appreciated. Maybe I have missed out on something?

Thanks

**Comment by amanjeetsingh** [comment]
[debug-squash.txt](https://github.com/goldmann/docker-squash/files/311109/debug-squash.txt)

**Comment by goldmann** [comment]
Thanks for the debug log, I'll take a look at this and try to create a reproducer. You are right about the loading phase causing the failure. This means that Docker expects to have this file present, but we hide it with whiteout file (or other way around). I'm on PTO for this week, so I may not be able to sig into this immediately, sorry.

I see you use some development version, probably installed from master, can you let me know which commit it was?

**Comment by amanjeetsingh** [comment]
Hi, 

Thanks heaps for looking into this. 

I cloned master branch last week (around 8th June 2016)  and going by commit history on Git, I would say that it is from commit 382ff9907d24247d6ab81cfc3f00bf342a633e3a (which was the last commit to Master back on 25th May 2016).

**Comment by amanjeetsingh** [comment]
Tried version 1.0.0rc6 which I downloaded using 
pip -v install docker-squash --pre

Same big image, same error. (see attachment). 

Tried base CentOS image to see if docker-squash fails (thought there might be some issues around handling yum directories in CentOS, but docker-squash worked fine with this modified CentOS image. I made changes to the image before committing it to local repo so that there were a few layers for the program to squash. Docker-squash worked fine with this modified image. No linking errors were reported. 

[Uploading error-log.zip…]()

**Comment by amanjeetsingh** [comment]
[error-log.zip](https://github.com/goldmann/docker-squash/files/318306/error-log.zip)

**Comment by goldmann** [comment]
OK, I'm back from PTO and can look at it in more details. Stay tuned.

**Comment by goldmann** [comment]
Any chance you could run the build again with code from `whiteout-fix` branch? This fixes some other issue in the area where I would expect to have a bug causing your failure. It's worth to give a try.

```
pip install -U --user https://github.com/goldmann/docker-squash/archive/whiteout-fix.zip
```

**Comment by amanjeetsingh** [comment]
Sure. I will test it out and update this ticket.

Thanks for looking into this.

**Comment by amanjeetsingh** [comment]
[docker-squash-whiteoutFix-error.zip](https://github.com/goldmann/docker-squash/files/328873/docker-squash-whiteoutFix-error.zip)

**Comment by amanjeetsingh** [comment]
Tried "whiteout-fix" branch and it failed again with a similar error while trying to link. 

2016-06-22 22:55:50,446 root         ERROR    404 Client Error: Not Found ("Untar re-exec error: exit status 1: output: link /var/lib/yum/yumdb/b/143287d0e91920a2d60b71cc8c0e48ae5d0627e4-bind-utils-9.8.2-0.30.rc1.el6-x86_64/reason /var/lib/yum/yumdb/g/878780cd2c05575bec8dc0bc343755f16a888c82-grub-0.97-93.el6-x86_64/reason: no such file or directory")

I have uploaded full log file as well. "docker-squash-whiteoutFix-error.zip"

**Comment by goldmann** [comment]
It was worth a try at least, thanks for testing.

**Comment by goldmann** [comment]
I do see you use a EL6 base image, can you tell me what image it is? Is it a custom one? I have a feeling that this is a pretty old one. Are you doing full system update there (`yum -y update`)?

This issue can be related some some special files (hard links maybe? this is suggested by: `Untar re-exec error: exit status 1: output: link /var/lib/yum/yumdb/b/143287d0e91920a2d60b71cc8c0e48ae5d0627e4-bind-utils-9.8.2-0.30.rc1.el6-x86_64/reason`). But I thought I fixed it some time ago for all special files, not only hard links, as part of #82, as you mentioned.

I wonder how accurate is the mention of file in the error message. For example if I focus on the bind-utils file, I do see in my local tests where I upgrade EL 6.6 image using `yum update` call following message:

```
2016-06-23 12:59:55,710 root         DEBUG    Adding 'var/lib/yum/yumdb/b/.wh.143287d0e91920a2d60b71cc8c0e48ae5d0627e4-bind-utils-9.8.2-0.30.rc1.el6-x86_64' marker file back...
```

I do not see this in your log which means that because of some reason it was, but it possibly should be there.

Hard to find the cause of this issue without a smaller reproducer, tbh. Can you maybe strip down your Dockerfile to just use FROM and do changes that involve packages (`yum`, `rpm`) test and share it with me if it fails for you?

**Comment by goldmann** [comment]
OK, I could reproduce it locally, stay tuned! :)

**Comment by goldmann** [comment]
The reason it fails for you is that you squash all layers and additionally there are some hard link there in the lowest layers. This is an edge case. I'll fix it, but in general - avoid squashing all layers - best practice it to squash content of a single Dockerfile.

**Comment by goldmann** [comment]
It seems I have a dirty fix for this, even integs are green. Weird :)

Need to refactor it though - it's a bit too verbose now.

**Comment by goldmann** [comment]
OK, could you please give now `404-error` branch a try?

```
pip install -U --user https://github.com/goldmann/docker-squash/archive/404-error.zip
```

**Comment by amanjeetsingh** [comment]
Thanks. Will post my feedback later today. 

In regards to the image - one of the other engineers in my team created the image. I do recall him mentioning that he downloaded a Centos6 - Greenplum single node Docker image from Git (this was 6 months ago). We then modified the image so that Greenplum database starts up automatically when we start a container.  I am waiting on him to send me the Dockerfile he used to create this image. 

The only reason I am squashing all layers is that since we went LIVE with this image at our workplace, developers have applied lots of changes to it via our CI process and over a period of time, it has grown to over 50 GB and now it takes a while to work with this large image. Hence, I am trying to squash all the layers in one go. 

Idea is that if we can get docker-squash to work, we will use it to flatten the image at least once a week. In which case, it will only be flattening a limited number (5-6) layers, instead of 20+ it's trying to do now. Hope it makes sense. 

Thanks once again for looking into this!

**Comment by goldmann** [comment]
Yes, looks like a plan for me :)

Please let me know if this #102 solves your issue. I'm planning to cut a final release after this issue is fixed.

**Comment by amanjeetsingh** [comment]
Hi, 
I gave it a go earlier today. Definitely no "hard link" errors but this time around, I am getting errors because hash values for a layer doesn't match with what the squashing program expects. Attached are the output logs from two runs. In the first test, I tried squashing all layers and in the second test, I tried squashing 4 layers. 


Error - 

2016-06-27 13:29:13,370 root         DEBUG    Cleaning up /tmp/docker-squash-pXbjY4 temporary directory
2016-06-27 13:29:15,099 root         ERROR    500 Server Error: Internal Server Error ("invalid diffID for layer 15: expected "sha256:41502091c1af09843be9c9a5183dc68d645f285783f573148807633bdef47e49", got "sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"")

**Comment by amanjeetsingh** [comment]
[error-squash4Layers.log.gz](https://github.com/goldmann/docker-squash/files/338843/error-squash4Layers.log.gz)
[error-squashAllLayers.log.gz](https://github.com/goldmann/docker-squash/files/338844/error-squashAllLayers.log.gz)

**Comment by amanjeetsingh** [comment]
Since the size of the image was getting out of hand (55 GB+) and slowing down our CD process, I had to go back to our Greenplum single node image (approx 9 GB) and load production Greenplum objects manually. Size is now back to 11 GB. This should work for us for next couple of weeks before it balloons to 50 GB+ in 2-3 weeks' time! Don't have that big image anymore. I did try running docker-squash against this smaller image but I was still getting hash value mismatch errors (like the ones captured in the error log above).

**Comment by goldmann** [comment]
OK, this is new for me, but at least we're moving forward.

I'm wondering how should I reproduce it now :) But this looks like something trivial. Obviously finding it will take some time I suppose.

BTW, I haven't tested `docker-squash` on Docker 1.11.1 so you are probably the first one. I guess it shouldn't really matter that you use newer version.

**Comment by amanjeetsingh** [comment]
From the errors that I am getting now, I would say that there's something "funny" with the image. 
Definitely not getting any "link" errors. 

I just re-ran the docker-squash against Docker image with our production metadata, and it errror'd out with the following error. Here's an excerpt, 

2016-06-29 19:36:22,700 root         DEBUG    Found a hard link var/lib/yum/yumdb/g/878780cd2c05575bec8dc0bc343755f16a888c82-grub-0.97-93.el6-x86_64/reason to a file which is marked to be skipped: var/lib/yum/yumdb/b/143287d0e91920a2d60b71cc8c0e48ae5d0627e4-bind-utils-9.8.2-0.30.rc1.el6-x86_64/reason, skipping hard link too
2016-06-29 19:36:24,448 root         DEBUG    Found a hard link var/lib/yum/yumdb/r/561d9d2308869c7178a1ad5a7b5a3cc8aec0e24d-rootfiles-8.1-6.1.el6-noarch/reason to a file which is marked to be skipped: var/lib/yum/yumdb/b/143287d0e91920a2d60b71cc8c0e48ae5d0627e4-bind-utils-9.8.2-0.30.rc1.el6-x86_64/reason, skipping hard link too
2016-06-29 19:36:24,471 root         DEBUG    Found a hard link var/lib/yum/yumdb/r/9e293d16a69eea046d36afeacc9433b547534ab8-redhat-logos-60.0.14-12.el6.centos-noarch/from_repo to a file which is marked to be skipped: var/lib/yum/yumdb/M/1b10c8c9eb7cfd86594e226ca4f00e56aa1ac2c5-MAKEDEV-3.24-6.el6-x86_64/from_repo, skipping hard link too
2016-06-29 19:36:24,475 root         DEBUG    Found a hard link var/lib/yum/yumdb/r/9e293d16a69eea046d36afeacc9433b547534ab8-redhat-logos-60.0.14-12.el6.centos-noarch/from_repo_revision to a file which is marked to be skipped: var/lib/yum/yumdb/M/1b10c8c9eb7cfd86594e226ca4f00e56aa1ac2c5-MAKEDEV-3.24-6.el6-x86_64/from_repo_revision, skipping hard link too
2016-06-29 19:36:24,479 root         DEBUG    Found a hard link var/lib/yum/yumdb/r/9e293d16a69eea046d36afeacc9433b547534ab8-redhat-logos-60.0.14-12.el6.centos-noarch/from_repo_timestamp to a file which is marked to be skipped: var/lib/yum/yumdb/M/1b10c8c9eb7cfd86594e226ca4f00e56aa1ac2c5-MAKEDEV-3.24-6.el6-x86_64/from_repo_timestamp, skipping hard link too
2016-06-29 19:36:24,483 root         DEBUG    Found a hard link var/lib/yum/yumdb/r/9e293d16a69eea046d36afeacc9433b547534ab8-redhat-logos-60.0.14-12.el6.centos-noarch/installed_by to a file which is marked to be skipped: var/lib/yum/yumdb/M/1b10c8c9eb7cfd86594e226ca4f00e56aa1ac2c5-MAKEDEV-3.24-6.el6-x86_64/installed_by, skipping hard link too
2016-06-29 19:36:24,496 root         DEBUG    Found a hard link var/lib/yum/yumdb/r/9e293d16a69eea046d36afeacc9433b547534ab8-redhat-logos-60.0.14-12.el6.centos-noarch/releasever to a file which is marked to be skipped: var/lib/yum/yumdb/M/1b10c8c9eb7cfd86594e226ca4f00e56aa1ac2c5-MAKEDEV-3.24-6.el6-x86_64/releasever, skipping hard link too
2016-06-29 19:36:24,953 root         INFO     Squashing finished!
2016-06-29 19:45:35,481 root         INFO     New squashed image ID is e1b505a9889a182992445173305aa2bbe7c32aab8082e639d41bcff6fa191836
2016-06-29 19:45:35,505 root         DEBUG    Generating tar archive for the squashed image...
2016-06-29 19:53:19,114 root         DEBUG    Archive generated
2016-06-29 19:53:19,114 root         DEBUG    Loading squashed image...
2016-06-29 20:00:43,926 requests.packages.urllib3.connectionpool DEBUG    "POST /v1.23/images/load HTTP/1.1" 404 146
2016-06-29 20:00:44,145 root         DEBUG    Cleaning up /tmp/docker-squash-25yfXN temporary directory
2016-06-29 20:00:45,787 root         ERROR    404 Client Error: Not Found ("open /var/lib/docker/tmp/docker-import-890183688/e1b505a9889a182992445173305aa2bbe7c32aab8082e639d41bcff6fa191836.json: no such file or directory")
Traceback (most recent call last):
  File "/root/.local/lib/python2.7/site-packages/docker_squash/cli.py", line 62, in run
    from_layer=args.from_layer, tag=args.tag, output_path=args.output_path, tmp_dir=args.tmp_dir, development=args.development).run()
  File "/root/.local/lib/python2.7/site-packages/docker_squash/squash.py", line 58, in run
    return self.squash(image)
  File "/root/.local/lib/python2.7/site-packages/docker_squash/squash.py", line 81, in squash
    image.load_squashed_image()
  File "/root/.local/lib/python2.7/site-packages/docker_squash/image.py", line 231, in load_squashed_image
    self._load_image(self.new_image_dir)
  File "/root/.local/lib/python2.7/site-packages/docker_squash/image.py", line 275, in _load_image
    self.docker.load_image(f)
  File "/usr/local/lib/python2.7/dist-packages/docker/api/image.py", line 148, in load_image
    self._raise_for_status(res)
  File "/usr/local/lib/python2.7/dist-packages/docker/client.py", line 153, in _raise_for_status
    raise errors.NotFound(e, response, explanation=explanation)
NotFound: 404 Client Error: Not Found ("open /var/lib/docker/tmp/docker-import-890183688/e1b505a9889a182992445173305aa2bbe7c32aab8082e639d41bcff6fa191836.json: no such file or directory")
2016-06-29 20:00:46,143 root         ERROR    Execution failed, consult logs above. If you think this is our fault, please file an issue: https://github.com/goldmann/docker-squash/issues, thanks!

**Comment by goldmann** [comment]
It looks like the metadata file `e1b505a9889a182992445173305aa2bbe7c32aab8082e639d41bcff6fa191836.json` was not created. Which would be totally weird for me.

**Comment by goldmann** [comment]
Any chance you could share the last 4 instructions in the Dockerfile? It would be perfect if you could share the image with me (privately). I cannot reproduce this failure. I tested with Docker 1.11.0 and 1.11.2, everything works fine for an image created from this Dockerfile:

```
FROM centos:6.6
RUN yum clean all
RUN yum -y update
```

@amanjeetsingh - can you please help me with this?

**Comment by amanjeetsingh** [comment]
Hi,
Sorry, cannot share the image as is for obvious reasons. However, I can sanitise it and then share it with you. Will take me a day or two. 

In the meanwhile, I have attached output of "docker history". I'll speak to the engineering team and try to get hold of the Dockerfile. 

Thanks

**Comment by amanjeetsingh** [comment]
[docker-history.txt](https://github.com/goldmann/docker-squash/files/341339/docker-history.txt)

**Comment by goldmann** [comment]
Sure, I understand this :) Thanks for the history, at least this give me a hint what's going on. It seems that most fun one is the latest layer. Unfortunately I'll not be able to create something similar and put it into our integration test suite :(

I see multiple places where the Dockerfile could be optimized. You probably should spend some time on finding these (for example run `ENV` instructions in one call, externalize complex instructions to bash scripts, etc).

**Comment by amanjeetsingh** [comment]
Hi,

Just created a new repo with the Dockerfiles for creating Greenplum Docker image. I have removed some bits of the code in order to sanitise it. Let me know how you go. Thanks.

https://github.com/amanjeetsingh/GP-docker

**Comment by goldmann** [comment]
I looked at the repo but unfortunately I cannot do with it anything. I'm missing a lot of thing to be able to build it, but you will not be able to provide them to me. We have a deadlock here :)

**Comment by goldmann** [comment]
Oh, one sec, there is something on the HUB. Checking...

**Comment by goldmann** [comment]
I've tried the image on HUB and I was able to squash it without issues.

**Comment by goldmann** [comment]
I merged the PR #102 which is supposed to solve the main issue. When you'll have a reproducer for the other issue, please open a new ticket and I'll fix it. Unfortunately I cannot do it now :(

**Comment by amanjeetsingh** [comment]
Thanks for looking into this. As I said earlier, "hard links" error has been resolved. Will see how we go with a fresh image that we are working on. 
Thanks!
